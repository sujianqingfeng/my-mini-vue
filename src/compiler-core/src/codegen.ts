import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

/**
 * 生成代码
 *
 * @param ast
 * @returns
 */
export function generate(ast) {
  const context = createCodegenContext(ast);
  const { push } = context;

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(",");

  push(`function ${functionName}(${signature}) {`);
  push("return ");

  genNode(ast.codegenNode, context);

  push("}");

  return {
    code: context.code,
  };
}

/**
 * 添加导入方法的代码
 *
 * @param ast
 * @param context
 */
function genFunctionPreamble(ast, context) {
  // 生成类似 const { toDisplayString: _toDisplayString } = Vue
  const { push } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;

  if (ast.helpers.length) {
    push(`const { ${ast.helpers.map(aliasHelper).join(",")} } = ${VueBinging}`);
  }

  push("\n");
  push("return ");
}

/**
 * 生成节点
 *
 * @param ast
 * @param context
 */
function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;

    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;

    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;

    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;

    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;

    default:
      break;
  }
}

/**
 * 生成复合类型的代码
 *
 *
 * @param node
 * @param context
 */
function genCompoundExpression(node, context) {
  const { children } = node;
  const { push } = context;

  for (let i = 0; i < children.length; i++) {
    const child = children[i];

    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function genElement(node, context) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);

  const nullableNodes = genNullable([tag, props, children]);
  genNodeList(nullableNodes, context);

  push(")");
}

/**
 * 处理多个节点
 *
 * @param nodes
 * @param context
 */
function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }

    if (i < nodes.length - 1) {
      push(",");
    }
  }
}

/**
 * 处理虚值
 *
 * @param args
 * @returns
 */
function genNullable(args) {
  return args.map((arg) => arg || "null");
}

/**
 * 处理插值中表达式
 *
 * @param node
 * @param context
 */
function genExpression(node, context) {
  const { push } = context;
  push(node.content);
}

/**
 * 处理插值
 *
 * @param node
 * @param context
 */
function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

/**
 * 处理文本节点
 *
 * @param node
 * @param context
 */

function genText(node: any, context: any) {
  const { push } = context;
  push(`"${node.content}"`);
}

/**
 * 生成上下文 存储生成的代码已经工具函数
 *
 * @param ast
 * @returns
 */
function createCodegenContext(ast: any) {
  const context = {
    code: "",
    push: (source) => {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}
