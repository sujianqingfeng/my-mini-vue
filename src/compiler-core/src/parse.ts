import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context = createParseContext(content);
  return createRoot(parseChildren(context, []));
}

/**
 * 创建解析的上下文 按照一定格式存储数据
 *
 * @param content
 * @returns
 */
function createParseContext(content) {
  return {
    source: content,
  };
}

/**
 * 解析内容为子节点
 *
 * @param context
 * @returns
 */
function parseChildren(context, ancestors) {
  const nodes: any = [];

  while (!isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    if (s.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (s[0] === "<") {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors);
      }
    }

    if (!node) {
      node = parseText(context);
    }

    nodes.push(node);
  }
  return nodes;
}

/**
 * children 解析完成标志
 *
 * @param context
 * @returns
 */
function isEnd(context, ancestors) {
  const s = context.source;

  if (s.startsWith("</")) {
    for (let i = 0; i < ancestors.length; i++) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }

  return !s;
}

/**
 * 解析text节点
 * * @param context
 * @returns
 */
function parseText(context: any) {
  let endIndex = context.source.length;
  const endTokens = ["{{", "<"];

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);

    if (index != -1 && index < endIndex) {
      endIndex = index;
    }
  }

  const content = parseTextData(context, endIndex);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

/**
 * 解析text数据
 *
 * @param context
 * @param length
 * @returns
 */
function parseTextData(context: any, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}

/**
 * 解析element
 *
 * @param context
 * @returns
 */
function parseElement(context: any, ancestors) {
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  // 注意顺序是在endTag之前
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    throw new Error(`unclosed tag ${element.tag}`);
  }

  return element;
}

/**
 * 解析tag
 *
 * @param context
 * @param type
 * @returns
 */
function parseTag(context: any, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  let tag = "";
  tag = match[1];
  advanceBy(context, match[0].length);
  advanceBy(context, 1);

  if (type === TagType.End) {
    return;
  }

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

/**
 * 解析插值
 *
 * @param context
 * @returns
 */
function parseInterpolation(context) {
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  advanceBy(context, openDelimiter.length);
  const rowContentLength = closeIndex - openDelimiter.length;

  const rowContent = parseTextData(context, rowContentLength);

  const content = rowContent.trim();

  advanceBy(context, closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

/**
 * 截取内容
 *
 * @param context
 * @param length
 */
function advanceBy(context, length) {
  context.source = context.source.slice(length);
}

/**
 * 创建根节点
 *
 * @param children
 * @returns
 */
function createRoot(children) {
  return { children, type: NodeTypes.ROOT };
}

/**
 * 判断文本开头是否是传入的tag
 *
 * @param source
 * @param tag
 */
function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}
