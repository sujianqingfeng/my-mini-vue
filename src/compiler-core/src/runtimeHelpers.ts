export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREATE_ELEMENT_VNODE = Symbol("createElementVNode");

/**
 * 函数名映射
 *
 */
export const helperMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [CREATE_ELEMENT_VNODE]: "createElementVNode",
};
