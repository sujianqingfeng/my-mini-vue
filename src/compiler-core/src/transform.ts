export function transform(root, options = {}) {
  const context = createTransformContext(root, options);

  traverseNode(root, context);

  createRootCodegen(root);
}

/**
 * 创建生成code的根节点
 *
 * @param root
 */
function createRootCodegen(root: any) {
  root.codegenNode = root.children[0];
}

/**
 * 便利节点
 *
 * @param node
 * @param context
 */
function traverseNode(node: any, context) {
  // 执行处理函数
  const { nodeTransforms } = context;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(node);
  }

  // 遍历子节点
  traverseChildren(node, context);
}

/**
 * 便利子节点
 *
 * @param node
 * @param context
 */
function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      traverseNode(children[i], context);
    }
  }
}

/**
 * 创建transform context
 *
 * @param root
 * @param options
 * @returns
 */
function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };

  return context;
}
