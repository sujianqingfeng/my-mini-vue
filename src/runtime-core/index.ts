export * from "../reactivity";
export { createRenderer } from "./renderer";
export { createApp } from "../runtime-dom";
export { renderSlots } from "./helps/renderSlots";
export { createTextVNode, createElementVNode } from "./vnode";
export { getCurrentInstance, registerRuntimeCompile } from "./component";
export { h } from "./h";
export { provide, inject } from "./apiInject";
export { nextTick } from "./scheduler";
export { toDisplayString } from "../shared";
