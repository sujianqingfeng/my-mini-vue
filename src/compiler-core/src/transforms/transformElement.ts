import { createVNodeCall, NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const { children, tag } = node;

      const vnodeTag = `"${tag}"`;
      const vnodeProps = undefined;
      const vnodeChildren = children[0];

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
