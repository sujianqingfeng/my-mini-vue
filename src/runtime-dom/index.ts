import { createRenderer } from "../runtime-core";
import { isOn } from "./../shared/index";

/**
 * 创建节点
 *
 * @param type
 * @returns
 */
function createElement(type) {
  return document.createElement(type);
}

/**
 * 设置prop
 *
 * @param el
 * @param key
 * @param val
 */
function patchProp(el, key, val) {
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase();
    el.addEventListener(event, val);
  } else {
    el.setAttribute(key, val);
  }
}

/**
 * 添加节点
 *
 * @param el
 * @param parent
 */
function insert(el, parent) {
  parent.append(el);
}

// TODO 更多的dom操作方法

const renderer: any = createRenderer({ createElement, patchProp, insert });

export function createApp(...args) {
  // 这里有点绕  这里调用的实际是crateApp.ts中createAppAPI中的createApp
  // renderer = { createApp: createAppAPI = { createApp }  }
  return renderer.createApp(...args);
}

export * from "../runtime-core";
