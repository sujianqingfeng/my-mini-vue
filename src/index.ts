export * from "./runtime-dom";
import { registerRuntimeCompile } from "./runtime-dom";

import { baseCompile } from "./compiler-core/src";
import * as runtimeDom from "./runtime-dom";

export function compileToFunction(template) {
  const { code } = baseCompile(template);
  // 这里记得要调用
  const render = new Function("Vue", code)(runtimeDom);

  return render;
}

registerRuntimeCompile(compileToFunction);
