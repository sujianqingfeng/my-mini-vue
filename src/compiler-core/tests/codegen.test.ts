import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { generate } from "../src/codegen";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");

    transform(ast);
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");

    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div></div>");

    transform(ast, {
      nodeTransforms: [transformElement],
    });
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("element union", () => {
    const ast = baseParse("<div>hi,{{message}}</div>");

    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });
});
