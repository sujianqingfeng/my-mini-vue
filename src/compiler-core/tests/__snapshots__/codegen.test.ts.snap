// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache) {return _createElementVNode(div)}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx,_cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx,_cache) {return \\"hi\\"}"
`;
